### PR TITLE
Provide DataFrame API for `map` and move `map` to `functions-array`

### LINKED ISSUE #11546: Easier Dataframe API for `map`
Dataframe API for `map` expects us to pass args with `make_array`

i.e.
`map(vec![make_array(vec![lit("a"), lit("b")]), make_array(vec![lit("1"), lit("2")])])

I think we could have easier one with without `make_array`
`map(vec![lit("a"), lit("b")], vec![lit("1"), lit("2")]])

To achieve this we may need to change the arguments o f `MapFunc` from two array to `Vec<Expr>`, which the first half are `keys`, another half are `values`.

_Originally posted by @jayzhan211 in https://github.com/apache/datafusion/pull/11452#discussion_r1684115054_

### PR DESCRIPTION
**Pull request by goldmedal**
## Which issue does this PR close?

Closes #11546 .

## What changes are included in this PR?
- Add a Dataframe API `map()`.
- Because some implementation of `map` function is based on `array`, I moved `map` to `functions-array`. See more discussion in https://github.com/apache/datafusion/issues/11546#issuecomment-2241453265

## Are these changes tested?
yes

## Are there any user-facing changes?
Add a function

### CODE REVIEW

File: datafusion/core/tests/dataframe/dataframe_functions.rs [hunk 1]
```diff
@@ -1087,3 +1088,20 @@ async fn test_fn_array_to_string() -> Result<()> {
 
     Ok(())
 }
+
+#[tokio::test]
+async fn test_fn_map() -> Result<()> {
+    let expr = map(vec![lit("a"), lit("b"), lit("c")], vec![lit(1), lit(2), lit(3)]);
```
Comment (goldmedal) [inline datafusion/core/tests/dataframe/dataframe_functions.rs:1094]: I expected the DataFrame user can use `map` like this.

**Comment by jayzhan211** [comment]
Can we also add `map` to `roundtrip_expr_api`?

**Review by jayzhan211 — APPROVED** [review]
👍

**Review by alamb — APPROVED** [review]
I agree map makes sense to be in the functions-array crate. Thank you @goldmedal and @jayzhan211

**Comment by goldmedal** [comment]
Thanks @jayzhan211 and @alamb for reviewing!

**Pull request merged**